### DISCUSSION TITLE
What features would an awesome Config Builder have?

### DISCUSSION THREAD

**Discussion by kmcq** [open-truss/open-truss#161] [Ideas]
Hey @open-truss/engineers –

I think the Config Builder is a project-defining feature of Open Truss and will be investing some time in making it awesome. I wanted to see what ideas you all have and am opening this discussion to hear 'em. The sky's the limit, no idea is too big!

I've listed some of my own ideas; feel free to comment on them to show support and share ideas. I'll certainly do the same on any that get added by others.

**Comment by kmcq** [comment]
You should be able to edit a component's props in-line. By this I mean there's a floating element on each frame where you can edit all aspects of it.

**Comment by kmcq** [comment]
The Config Builder needs a prop for exporting configs.

**Reply by jonmagic** [reply]
Would this be in addition to saving?

**Reply by kmcq** [reply]
Export is my generic way of saying saving 👍 A simple version could just put it in your paste buffer

**Comment by kmcq** [comment]
You should be able to reorder frames (without deleting and re-adding). In addition, frames should be able to be un-nested. Put together, frames should be able to be easily moved from anywhere in the config to anywhere else.

**Comment by kmcq** [comment]
Editing the generated YAML directly should be possible and as good as editing YAML in a good text editor.

**Comment by andrew-dillon** [comment]
The generated YAML should be elegant and readable, as if written by a human's hands. It shouldn't be obvious that it was generated by a computer.

**Comment by hktouw** [comment]
To make collaborating on the config builder easier given so much of it is UI heavy,  @kmcq and I were talking about how it could make sense to put the entire config builder into the package instead of just hooks. We would have to commit to an OT branded UI / look, but that is probably fine since it is an OT specific page. In order for this to work, we'll have to research what a good dev setup might be so that it is easy to import the config builder into apps and have all the assets compile accordingly.

@jonmagic I know you preferred to not go down the approach of having the package have UI, but it will likely become tedious to copy UI code back and forth between the demo app / our app / Keith's and fix breaking changes. Do you have thoughts on how we can make the dev experience easier for building the config builder?

**Reply by jonmagic** [reply]
If you can pass in `style` and `class` aka `className` props to each of the components that make up the config builder then we can probably style things in individual apps.

**Reply by hktouw** [reply]
I think that still may have a similar issue where we would have to manage three separate styles between our apps and ensure new changes don't break each other. That still feels a tad brittle. 

What if we agreed on adding UI framework (shadcn?) to each of our apps? That way we can build the config builder using one framework / set of styles, import the package to our apps, and our apps should be able to build the project accordingly.

**Reply by jonmagic** [reply]
You'd actually be adding tailwind in that case, and I'm not sure that's gonna fly.

**Reply by kmcq** [reply]
I'm asking around what a good way of doing this. Perhaps naive but we could define all the styles in `style` props and have the components take in an object that can override those: `style={{ color: 'blue', ...styles }}`.

**Reply by hktouw** [reply]
That means we can't really use classes or utility classes, but maybe that is not too awful. style only development like it's the year 2000. :)

Another option is we could setup our own stylesheets (or use tailwind) in the package and generate a minified css file to include in the project. Then all apps need to do is import that stylesheet and things should just work. There are probably some local environment details we'd need to hammer out though to make sure the dev experience is good loading the config builder in your app and having assets compile in the package.

**Reply by kmcq** [reply]
I thought some more about this and I agree: having OT export a CSS file is the way to go. This means consumers could easily just not use that file and write CSS for the same classes to get things to look how they want

**Reply by hktouw** [reply]
I was a little worried about class name conflicts across stylesheets, but I think we can use the [`@scope`](https://developer.mozilla.org/en-US/docs/Web/CSS/:scope) pseudo attribute to scope our stylesheets to within anything with a `open-truss` class. From the Mozilla docs

```html
<div class="light-scheme">
  <p>
    MDN contains lots of information about
    <a href="/en-US/docs/Web/HTML">HTML</a>,
    <a href="/en-US/docs/Web/CSS">CSS</a>, and
    <a href="/en-US/docs/Web/JavaScript">JavaScript</a>.
  </p>
</div>

<div class="dark-scheme">
  <p>
    MDN contains lots of information about
    <a href="/en-US/docs/Web/HTML">HTML</a>,
    <a href="/en-US/docs/Web/CSS">CSS</a>, and
    <a href="/en-US/docs/Web/JavaScript">JavaScript</a>.
  </p>
</div>
```

```css
@scope (.light-scheme) {
  :scope {
    background-color: plum;
  }

  a {
    color: darkmagenta;
  }
}

@scope (.dark-scheme) {
  :scope {
    background-color: darkmagenta;
    color: antiquewhite;
  }

  a {
    color: plum;
  }
}
```

![:scope - CSS: Cascading Style Sheets MDN 2024-04-26 at 9 29 31 AM](https://github.com/open-truss/open-truss/assets/4481584/551ce8c2-41ab-4b0b-983f-d8129e41065a)

**Comment by spicycode** [comment]
I think one of the things I really like about good config builders is how easily they highlight invalid configurations.  I'm thinking about this through the lens of migrating configuration files from version to version in developing systems.

**Reply by spicycode** [reply]
Imagine you upgrade from a beta to v1.0, and it says "Hey I noticed your configuration had some issues.  We've auto-migrated it to be compatible with v1.0, but would like to request your review before putting it into action"

**Comment by hktouw** [comment]
I would like to see the config builder support solid documentation for each component. e.g. what each prop in the component is used for, how to use it, and links to other workflows that use this. This would be quite useful for complex components. e.g. `OTUqiDataprovider` is quite complex and not intuitive how to use. Some thoughts on this:
- I wonder if we want to adjust the OT component prop interface to not only allow setting the zod type, but also setting documentation. That might be one way how we can get documentation in a machine readable format we can then show in the config builder.
- In the short term, perhaps a link to a GitHub search is how we can link to configs using the component

**Comment by hktouw** [comment]
I think an important part of the config builder could be making component discovery easy. I wonder if there is a way we could have a storybook js like interface (or just use storybook) for this.

**Reply by kmcq** [reply]
Ooh borrowing (or using) storybook is a very cool idea. Storybook has places to write examples and documentation -- it would be great to support that.
